pkg/filesystem/subvolume.go

```diff
@@ -201,17 +203,107 @@ func unMarshaljson(list string) []fsStruct {
 	if errg != nil {
 		logging.Fatal(errg)
 	}
-
 	return unmarshal
 }
 
 func Delete(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespace, CephClusterNamespace, fs, subvol, svg string) {
 	k8sSubvolume := getK8sRefSubvolume(ctx, clientsets)
 	_, check := k8sSubvolume[subvol]
 	if !check {
-		exec.RunCommandInOperatorPod(ctx, clientsets, "ceph", []string{"fs", "subvolume", "rm", fs, subvol, svg, "--retain-snapshots"}, OperatorNamespace, CephClusterNamespace, true)
-		logging.Info("subvolume %q deleted", subvol)
+		deleteOmapForSubvolume(ctx, clientsets, OperatorNamespace, CephClusterNamespace, subvol, fs)
+		_, err := exec.RunCommandInOperatorPod(ctx, clientsets, "ceph", []string{"fs", "subvolume", "rm", fs, subvol, svg, "--retain-snapshots"}, OperatorNamespace, CephClusterNamespace, true)
+		if err != nil {
+			logging.Fatal(err, "failed to delete subvolume of %s/%s/%s", fs, svg, subvol)
+		} else {
+			logging.Info("subvolume %s/%s/%s deleted", fs, svg, subvol)
+		}
 	} else {
-		logging.Info("subvolume %q is not stale", subvol)
+		logging.Info("subvolume %s/%s/%s is not stale", fs, svg, subvol)
 	}
 }
+
+func getMetadataPoolName(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespace, CephClusterNamespace, fs string) (string, error) {
+	fsstruct, err := getFileSystem(ctx, clientsets, OperatorNamespace, CephClusterNamespace)
+	if err != nil {
+		return "", err
+	}
+
+	for _, pool := range fsstruct {
+		if pool.Name == fs {
+			return pool.MetadataName, nil
+		}
+	}
+	return "", fmt.Errorf("metadataPool not found for %q filesystem", fs)
+}
+
+// deleteOmap deletes omap object and key for the given subvolume.
+func deleteOmapForSubvolume(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespace, CephClusterNamespace, subVol, fs string) {
+	logging.Info("Deleting the omap object and key for subvolume %q", subVol)
+	omapkey := getOmapKey(ctx, clientsets, OperatorNamespace, CephClusterNamespace, subVol, fs)
+	omapval := getOmapVal(subVol)
+	poolName, err := getMetadataPoolName(ctx, clientsets, OperatorNamespace, CephClusterNamespace, fs)
+	if err != nil || poolName == "" {
+		logging.Fatal(fmt.Errorf("pool name not found: %q", err))
+	}
+	if omapval != "" {
+		// remove omap object.
+		_, err := exec.RunCommandInOperatorPod(ctx, clientsets, "rados", []string{"rm", omapval, "-p", poolName, "--namespace", "csi"}, OperatorNamespace, CephClusterNamespace, true)
+		if err != nil {
+			logging.Error(err, "failed to remove omap object for subvolume %q", subVol)
```

Comment:
```suggestion
			logging.Error(err, "failed to remove omap object for subvolume %q", subVol)
```
`logging.Error` will only log and not return and exit and since this method doesn't return anything do you think we should use `logging.Fatal`?